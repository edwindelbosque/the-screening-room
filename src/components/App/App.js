import React, { Component } from "react";
import { Route } from "react-router-dom";
import Nav from "../Nav/Nav";
import AccessModal from "../AccessModal/AccessModal";
import Container from "../Container/Container";
import SelectedMovie from "../SelectedMovie/SelectedMovie";
import Footer from "../Footer/Footer";
import {
  getMovies,
  getWallpapers,
  postFavorite,
  removeFavorite,
  getFavorites,
  getFavoriteMovies,
  getFavoriteWallpapers
} from "../../apiCalls/apiCalls";
import {
  setMovies,
  setWallpapers,
  setLoading,
  setError,
  addFavorite,
  setFavorites,
  setUser,
  setRandomWallpaper,
  setFavoriteWallpapers
} from "../../actions";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { PropTypes } from "prop-types";
import "./App.scss";

export class App extends Component {
  componentDidMount = async () => {
    const {
      setWallpapers,
      hasError,
      setRandomWallpaper,
      setLoading
    } = this.props;
    try {
      setLoading(true);
      const wallpapers = await getWallpapers();
      await this.loadMovieData();
      setWallpapers(wallpapers);
      setRandomWallpaper(wallpapers);
      setLoading(false);
    } catch ({ message }) {
      hasError(message);
      setLoading(false);
    }
  };

  loadMovieData = async () => {
    const { setMovies, setFavorites, hasError, user, setFavoriteWallpapers } = this.props;
    const fetchedMovies = await getMovies();

    if (user.id) {
      try {
        const favoriteMovieIds = await getFavorites(user.id);
        const favorites = await getFavoriteMovies(favoriteMovieIds.favorites);
        setFavoriteWallpapers(await getFavoriteWallpapers(favorites));
        console.log(favorites);
        const markedMovies = await this.markFavorites(fetchedMovies, favorites);
        setMovies(markedMovies);
        await setFavorites(favorites);
      } catch ({ message }) {
        hasError(message);
      }
    } else {
      setMovies(fetchedMovies);
    }
  };

  markFavorites = (movies, favorites) => {
    return movies.map(movie => {
      return favorites.find(
        favorite => favorite.movie_id === movie.movie_id
      )
        ? { ...movie, favorite: true }
        : { ...movie, favorite: false };
    });
  };

  updateFavorites = async (movie, isFavorite) => {
    const { user, addFavorite, setFavorites, hasError } = this.props;
    if (!isFavorite) {
      try {
        let favoritesData = await postFavorite(movie, user.id);
        addFavorite(favoritesData);
      } catch ({ message }) {
        hasError(message);
      }
    } else {
      try {
        await removeFavorite(movie.movie_id, user.id);
        let newFavorites = await getFavorites(user.id);
        setFavorites(newFavorites.favorites);
      } catch ({ message }) {
        hasError(message);
      }
    }
  };

  logoutCurrentUser = () => {
    const { setUser } = this.props;
    setUser({});
  };

  render() {
    return (
      <main className="main">
        <Nav logoutCurrentUser={this.logoutCurrentUser} />
        <Route
          path="/(movies|favorites|search)/:id"
          render={({ match }) => {
            const { movies, searchResults, favorites } = this.props;
            const movieDetails = [...movies, ...searchResults, ...favorites].find(
              movie => movie.movie_id === parseInt(match.params.id)
            );
            console.log(searchResults, movies, movieDetails);
            return (
              <SelectedMovie
                movieDetails={movieDetails}
                match={match}
                wallpapers={[...this.props.wallpapers, ...this.props.favoriteWallpapers]}
              />
            );
          }}
        />
        <Route
          path="/(|movies|signup|login)"
          render={() => (
            <Container
              movies={this.props.movies}
              type="movies"
              updateFavorites={this.updateFavorites}
            />
          )}
        />
        <Route
          path="/(login|signup)"
          render={() => <AccessModal loadMovieData={this.loadMovieData} />}
        />
        <Route
          path="/favorites"
          render={() => (
            <Container
              movies={this.props.favorites}
              type="favorites"
              updateFavorites={this.updateFavorites}
            />
          )}
        />
        <Route
          path="/search"
          render={() => (
            <Container
              movies={this.props.searchResults}
              type="search"
              updateFavorites={this.updateFavorites}
            />
          )}
        />
        <Footer />
      </main>
    );
  }
}

export const mapStateToProps = ({
  movies,
  wallpapers,
  hasError,
  isLoading,
  user,
  favorites,
  searchResults,
  favoriteWallpapers
}) => ({
  movies,
  wallpapers,
  hasError,
  isLoading,
  user,
  favorites,
  searchResults,
  favoriteWallpapers
});

export const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      setMovies,
      setWallpapers,
      setLoading,
      setError,
      addFavorite,
      setFavorites,
      setUser,
      setRandomWallpaper,
      setFavoriteWallpapers
    },
    dispatch
  );
};
export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  movies: PropTypes.array,
  wallpapers: PropTypes.array,
  hasError: PropTypes.bool,
  isLoading: PropTypes.bool,
  user: PropTypes.object,
  favorites: PropTypes.array,
  setMovies: PropTypes.func,
  setWallpapers: PropTypes.func,
  setLoading: PropTypes.func,
  setError: PropTypes.func,
  addFavorite: PropTypes.func,
  setFavorites: PropTypes.func,
  setUser: PropTypes.func,
  setRandomWallpaper: PropTypes.func
};
