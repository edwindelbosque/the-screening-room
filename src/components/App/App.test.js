import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import {
  setMovies,
  setWallpapers,
  setLoading,
  setError,
  addFavorite
} from '../../actions/index';

jest.mock('../../apiCalls/apiCalls.js');

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('mapStateToProps', () => {
  it('should return an object with', () => {
    let movies, wallpapers, errMsg, isLoading, user, favorites;
    const mockState = {
      movies,
      wallpapers,
      errMsg,
      isLoading,
      user,
      favorites
    };

    const expected = {
      movies,
      wallpapers,
      errMsg,
      isLoading,
      user,
      favorites
    };

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.fn();
  });

  it('calls dispatch with a setMovies action', () => {
    const actionToDispatch = setMovies([{}, {}, {}, {}, {}]);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setMovies([{}, {}, {}, {}, {}]);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with setWallpapers', () => {
    const actionToDispatch = setWallpapers([{}, {}, {}, {}, {}]);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setWallpapers([{}, {}, {}, {}, {}]);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with setLoading', () => {
    const actionToDispatch = setLoading(true);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setLoading(true);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with errMsg', () => {
    const errMsg = {
      message: 'message'
    };
    const actionToDispatch = setError(errMsg);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setError({ message: 'message' });

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with addFavorite', () => {
    const newFavorite = {
      id: 17,
      movie_id: 290859,
      overview:
        "More than two decades have passed since Sarah Connor prevented Judgment Day, changed the future, and re-wrote the fate of the human race. Dani Ramos is living a simple life in Mexico City with her brother and father when a highly advanced and deadly new Terminator – a Rev-9 – travels back through time to hunt and kill her. Dani's survival depends on her joining forces with two warriors: Grace, an enhanced super-soldier from the future, and a battle-hardened Sarah Connor. As the Rev-9 ruthlessly destroys everything and everyone in its path on the hunt for Dani, the three are led to a T-800 from Sarah’s past that may be their last best hope.",
      poster_path:
        'https://image.tmdb.org/t/p/original/vqzNJRH4YyquRiWxCCOH0aXggHI.jpg',
      release_date: '2019-11-01',
      title: 'Terminator: Dark Fate',
      user_id: 2,
      vote_average: '6.8'
    };
    const actionToDispatch = addFavorite(newFavorite);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addFavorite({
      id: 17,
      movie_id: 290859,
      overview:
        "More than two decades have passed since Sarah Connor prevented Judgment Day, changed the future, and re-wrote the fate of the human race. Dani Ramos is living a simple life in Mexico City with her brother and father when a highly advanced and deadly new Terminator – a Rev-9 – travels back through time to hunt and kill her. Dani's survival depends on her joining forces with two warriors: Grace, an enhanced super-soldier from the future, and a battle-hardened Sarah Connor. As the Rev-9 ruthlessly destroys everything and everyone in its path on the hunt for Dani, the three are led to a T-800 from Sarah’s past that may be their last best hope.",
      poster_path:
        'https://image.tmdb.org/t/p/original/vqzNJRH4YyquRiWxCCOH0aXggHI.jpg',
      release_date: '2019-11-01',
      title: 'Terminator: Dark Fate',
      user_id: 2,
      vote_average: '6.8'
    });

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
