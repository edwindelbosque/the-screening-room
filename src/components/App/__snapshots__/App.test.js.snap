// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<main
  className="main"
>
  <Connect(Nav)
    logoutCurrentUser={[Function]}
  />
  <Route
    path="/(movies|favorites|search)/:id"
    render={[Function]}
  />
  <Route
    path="/(|movies|signup|login)"
    render={[Function]}
  />
  <Route
    path="/(login|signup)"
    render={[Function]}
  />
  <Route
    path="/favorites"
    render={[Function]}
  />
  <Route
    path="/search"
    render={[Function]}
  />
  <Footer />
</main>
`;
